### PR TITLE
=htc #18121 enable UTF-8 decoding for HTTP header values

### LINKED ISSUE #18121: Allow parsing of UTF8 header values
@sirthias can you check what is needed to enable this? It may be a simple fix to [scanHeaderValue](https://github.com/akka/akka/blob/release-2.3-dev/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpHeaderParser.scala#L482).

This is needed so we can fix #16821.

### PR DESCRIPTION
**Pull request by sirthias**
If we detect a potential UTF-8 byte sequence we attempt decoding, otherwise we leave the bytes as they are.

Closes #18121.

/cc @jrudolph

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpHeaderParser.scala [hunk 1]
```diff
@@ -375,6 +376,30 @@ private[engine] final class HttpHeaderParser private (
...
    */
   def formatSizes: String = s"$nodeCount nodes, ${branchDataCount / 3} branchData rows, $valueCount values"
+
+  // helpers for UTF-8 decoding,
+  // since they are only accessed when an UTF8 byte sequence is actually hit and UTF-8 sequences in header values are
+  // rare these fields can be lazy, the overhead of the lazy access should be overcompensated for by the saved
+  // allocations in the majority of cases
+  private lazy val byteBuffer = ByteBuffer.allocate(4)
+  private lazy val charBuffer = CharBuffer.allocate(2)
+  private lazy val decoder = UTF8.newDecoder()
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpHeaderParser.scala]: Thanks for the comment, sounds good.

File: akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpHeaderParserSpec.scala [hunk 2]
```diff
@@ -150,6 +150,13 @@ class HttpHeaderParserSpec extends WordSpec with Matchers with BeforeAndAfterAll
       parseAndCache("Fancy: foo\tbar\r\nx")() shouldEqual RawHeader("Fancy", "foo bar")
     }
 
+    "parse and cache a header with UTF8 chars in the value" in new TestSetup() {
+      parseAndCache("2-UTF8-Bytes: árvíztűrő ütvefúrógép\r\nx")() shouldEqual RawHeader("2-UTF8-Bytes", "árvíztűrő ütvefúrógép")
```
Comment (drewhk) [inline akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpHeaderParserSpec.scala]: thanks ;)

**Comment by drewhk** [comment]
LGTM

**Comment by ktoso** [comment]
LGTM, thanks!

File: akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpHeaderParserSpec.scala (hunk 2 above)
Comment (sirthias) [inline akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpHeaderParserSpec.scala]: As I said, the flood-resistant hammerdrill is now an integral part of my UTF-8 vocabulary!
Thanks to you! :)

Comment (jrudolph) [inline akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpHeaderParserSpec.scala]: Nice, is this somehow related to [this one](http://terraria.gamepedia.com/Chlorophyte_Drill) ?

Comment (sirthias) [inline akka-http-core/src/test/scala/akka/http/impl/engine/parsing/HttpHeaderParserSpec.scala]: Interesting! What is "flood-resistant chlorophyte drill" in Hungarian, @drewhk? Maybe we can squeeze a few more funky chars in there?

**Pull request merged**